### ISSUE TITLE
[Architecture] Create Shared Training Loop Base Class

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2597]
## Problem

8+ model training scripts duplicate training loop structure, with each example containing 360-922 lines of mostly identical code.

## Location

**Duplicate training loops in:**
- `examples/alexnet-cifar10/train.mojo` (551 lines)
- `examples/resnet18-cifar10/train.mojo` (360 lines)
- `examples/vgg16-cifar10/train.mojo` (922 lines)
- `examples/googlenet-cifar10/train.mojo` (502 lines)
- `examples/lenet5-mnist/train.mojo`
- `examples/simple-mlp/train.mojo`
- Plus 2+ more examples

## Current Behavior

Each training script independently implements:
1. Training loop (epochs, batches)
2. Forward pass
3. Loss computation
4. Backward pass
5. Optimizer step
6. Metrics tracking
7. Checkpoint saving
8. Validation loop

**Common structure:**
```mojo
fn train():
    # Model initialization (differs)
    var model = AlexNet(...)  # Or ResNet18, VGG16, etc.
    
    # Training loop (identical across examples)
    for epoch in range(num_epochs):
        for batch in dataloader:
            var output = model.forward(batch.input)
            var loss = criterion(output, batch.target)
            loss.backward()
            optimizer.step()
            # ... metrics, logging
        
        # Validation (identical)
        # Checkpoint saving (identical)
```

## Impact

- **Severity**: HIGH
- **Code duplication**: 360-922 lines × 8 examples = ~4500+ lines of duplicate code
- **Maintenance burden**: Bug fixes must be applied to all training scripts
- **Inconsistency**: Each example may implement training loop slightly differently
- **Developer productivity**: Copy-paste errors when creating new examples

### Option 1: Extend Existing `shared/training/trainer.mojo`

The codebase already has `shared/training/trainer.mojo`. Check if it can be extended:

```mojo
struct Trainer:
    """Base class for training neural networks.
    
    Handles common training loop logic:
    - Epoch/batch iteration
    - Forward/backward passes
    - Optimization steps
    - Metrics tracking
    - Checkpointing
    """
    
    fn __init__(out self, model: Model, optimizer: Optimizer, criterion: Loss):
        self.model = model
        self.optimizer = optimizer
        self.criterion = criterion
    
    fn train(mut self, dataloader: DataLoader, num_epochs: Int, validation_loader: Optional[DataLoader] = None):
        """Run training loop.
        
        Args:
            dataloader: Training data
            num_epochs: Number of epochs
            validation_loader: Optional validation data
        """
        for epoch in range(num_epochs):
            self._train_epoch(dataloader, epoch)
            
            if validation_loader:
                self._validate_epoch(validation_loader, epoch)
    
    fn _train_epoch(mut self, dataloader: DataLoader, epoch: Int):
        """Train single epoch (can be overridden)."""
        for batch in dataloader:
            var loss = self._train_step(batch)
            # Logging, metrics
    
    fn _train_step(mut self, batch: Batch) -> Float32:
        """Single training step (can be customized)."""
        var output = self.model.forward(batch.input)
        var loss = self.criterion(output, batch.target)
        loss.backward()
        self.optimizer.step()
        return loss.item()
    
    fn _validate_epoch(mut self, dataloader: DataLoader, epoch: Int):
        """Validation loop."""
        # Validation logic
```

### Option 2: Create New Base Class

If `trainer.mojo` doesn't exist or is unsuitable, create new base:

```mojo
trait TrainableModel:
    """Interface for models that can be trained."""
    fn forward(self, input: ExTensor) -> ExTensor
    fn parameters(self) -> List[ExTensor]

fn train_model[M: TrainableModel](
    model: M,
    dataloader: DataLoader,
    optimizer: Optimizer,
    criterion: Loss,
    num_epochs: Int = 10,
    validation_loader: Optional[DataLoader] = None
):
    """Generic training function.
    
    Works with any model implementing TrainableModel trait.
    """
    # Shared training loop implementation
```

### Migration Strategy

1. **Phase 1**: Create/extend base trainer class
2. **Phase 2**: Migrate one example (e.g., LeNet-5) to use base class
3. **Phase 3**: Validate training works identically
4. **Phase 4**: Migrate remaining examples
5. **Phase 5**: Remove duplicate training code

### Example-Specific Code

Allow customization through:
- Model instantiation (always different)
- Custom training step (if needed)
- Custom metrics (if needed)
- Callbacks for logging, checkpointing

**Migrated example:**
```mojo
fn main():
    # Model-specific initialization
    var model = AlexNet(num_classes=10)
    var optimizer = Adam(model.parameters(), lr=0.001)
    var criterion = CrossEntropyLoss()
    var dataloader = load_cifar10(batch_size=32)
    
    # Shared training loop
    var trainer = Trainer(model, optimizer, criterion)
    trainer.train(dataloader, num_epochs=50)
```

Reduced from 551 lines → ~50 lines per example.

## Benefits

- **DRY**: Single source of truth for training logic
- **Consistency**: All examples train identically
- **Maintainability**: Fix once, applies to all examples
- **Extensibility**: Easy to add features (LR scheduling, early stopping, etc.)
- **Simplicity**: New examples require minimal code

## Related Issues

- Part of architecture consolidation initiative
- Reduces codebase size significantly (~4000+ lines saved)

**Comment by mvillmow** [comment]
❌ Implementation failed:

```
Implementation failed after 3 attempts: Exit code 1: You've hit your limit · resets 4pm (America/Los_Angeles)

```

---
*Automated by implement_issues.py at 2025-12-19 13:16:23*

**Comment by mvillmow** [comment]
## Implementation Started

Starting implementation of shared training loop base class to consolidate duplicate training boilerplate across 7 model examples.

### Approach
Creating **TrainingScriptUtilities** module with:
1. Unified configuration struct with argument parsing
2. Training loop orchestration (epochs, LR scheduling, evaluation, checkpointing)
3. Unified dataset loading for CIFAR-10/EMNIST

### Branch
Created feature branch: `2597-shared-training-loop`

### Expected Impact
- 60-70% code reduction per training script
- ~1500-2000 lines of duplicated boilerplate consolidated
- Model-specific backward passes remain in each example (architecture-specific)
